activerecord/lib/active_record/inheritance.rb

```diff
@@ -245,7 +245,8 @@ def find_sti_class(type_name)
 
         def type_condition(table = arel_table)
           sti_column = arel_attribute(inheritance_column, table)
-          sti_names  = ([self] + descendants).map(&:sti_name)
+          base_class.pluck(inheritance_column).compact.uniq.map(&:safe_constantize)
```

Comment:
A few things here IMO:

  - This code only make sense if called in autoloaded environment. It shouldn't be called in production
  - It should use `base_class.uniq.pluck` so that the duplicates are handled by the database rather than done in memory.

Reply:
Also, do we really need the `.map(&:safe_constantize)` bit. The `pluck` should be returning all the `sti_names` that are currently in the database, anyway. It looks like we can just assign the return value of the `pluck` to `sti_names` and be done.

Reply:
We need the `safe_constantize` because we actually want to load the constant and retrieve the descendants of a given class.

We can't just pass the results of the `pluck` to sti_names, because this would load all the types for any given STI even if they are not descendants

```
Wizard = Class.new(ApplicationRecord)
Foo = Class.new(Wizard)
Magician = Class.new(Wizard)
Sorceler = Class.new(Magician)
```

Magician.first
# If we passed the result of `pluck` to the sti_names, Magician.first could return a `Foo` although Foo is not a descandant of Magician
```

Reply:
Right, that makes sense. For some reason, I was thinking the reification of the returned values would automatically handle that, but the explanation makes it clear that it doesn't. Thanks!